helix-view/src/clipboard.rs

```diff
@@ -203,16 +286,34 @@ mod provider {
             }
         }
 
-        fn get_contents(&self) -> Result<String> {
-            let output = self
-                .get_cmd
-                .execute(None, true)?
-                .context("output is missing")?;
-            Ok(output)
+        fn get_contents(&self, clipboard_type: ClipboardType) -> Result<String> {
+            match clipboard_type {
+                ClipboardType::Clipboard => Ok(self
+                    .get_cmd
+                    .execute(None, true)?
+                    .context("output is missing")?),
+                ClipboardType::Selection => {
+                    if let Some(cmd) = &self.get_primary_cmd {
+                        return cmd.execute(None, true)?.context("output is missing");
+                    }
+
+                    Ok(self.selection_buf.clone())
+                }
+            }
         }
 
-        fn set_contents(&self, value: String) -> Result<()> {
-            self.set_cmd.execute(Some(&value), false).map(|_| ())
+        fn set_contents(&mut self, value: String, clipboard_type: ClipboardType) -> Result<()> {
+            match clipboard_type {
+                ClipboardType::Clipboard => self.set_cmd.execute(Some(&value), false).map(|_| ()),
+                ClipboardType::Selection => {
+                    if let Some(cmd) = &self.set_primary_cmd {
+                        return cmd.execute(Some(&value), false).map(|_| ());
+                    }
+
+                    self.selection_buf = value;
+                    Ok(())
+                }
+            }
         }
```

Comment:
You could extract out the `CommandConfig` and then call `execute` (same for `get_contents`):

```suggestion
        fn set_contents(&mut self, value: String, clipboard_type: ClipboardType) -> Result<()> {
            let cmd = match clipboard_type {
                ClipboardType::Clipboard => &self.set_cmd,
                ClipboardType::Selection => {
                    if let Some(cmd) = &self.set_primary_cmd {
                        cmd
                    } else {
                        self.selection_buf = value;
                        return Ok(());
                    }
                }
            };
            cmd.execute(Some(&value), false).map(|_| ())
        }

```